File: NI-DAQ/Examples/DotNET4.0/Analog Out/Generate Voltage/GenMultVoltUpdatesIntClk_DigStart/VB/FunctionGenerator.vb
Imports System
Imports NationalInstruments.DAQmx
Imports System.Diagnostics

Public Enum WaveformType
    SineWave = 0
End Enum

Public Class FunctionGenerator
    Public Sub New( _
        ByVal timingSubobject As Timing, _
        ByVal desiredFrequency As String, _
        ByVal samplesPerBuffer As String, _
        ByVal cyclesPerBuffer As String, _
        ByVal type As String, _
        ByVal amplitude As String)

        Dim t As New WaveformType
        t = WaveformType.SineWave
        If (type = "Sine Wave") Then
            t = WaveformType.SineWave
        Else
            Debug.Assert(False, "Invalid Waveform Type")
        End If

        Init( _
            timingSubobject, _
            Double.Parse(desiredFrequency), _
            Double.Parse(samplesPerBuffer), _
            Double.Parse(cyclesPerBuffer), _
            t, _
            Double.Parse(amplitude))
    End Sub

    Public Sub New( _
        ByVal timingSubobject As Timing, _
        ByVal desiredFrequency As Double, _
        ByVal samplesPerBuffer As Double, _
        ByVal cyclesPerBuffer As Double, _
        ByVal type As WaveformType, _
        ByVal amplitude As Double)

        Init( _
            timingSubobject, _
            desiredFrequency, _
            samplesPerBuffer, _
            cyclesPerBuffer, _
            type, _
            amplitude)
    End Sub

    Private Sub Init( _
        ByVal timingSubobject As Timing, _
        ByVal desiredFrequency As Double, _
        ByVal samplesPerBuffer As Double, _
        ByVal cyclesPerBuffer As Double, _
        ByVal type As WaveformType, _
        ByVal amplitude As Double)

        If (desiredFrequency <= 0) Then
            Throw New ArgumentOutOfRangeException("desiredFrequency", desiredFrequency, "This parameter must be a positive number")
        ElseIf (samplesPerBuffer <= 0) Then
            Throw New ArgumentOutOfRangeException("samplesPerBuffer", samplesPerBuffer, "This parameter must be a positive number")
        ElseIf (cyclesPerBuffer <= 0) Then
            Throw New ArgumentOutOfRangeException("cyclesPerBuffer", cyclesPerBuffer, "This parameter must be a positive number")
        End If

        ' First configure the Task timing parameters
        If (timingSubobject.SampleTimingType = SampleTimingType.OnDemand) Then
            timingSubobject.SampleTimingType = SampleTimingType.SampleClock
        End If

        _desiredSampleClockRate = (desiredFrequency * samplesPerBuffer) / cyclesPerBuffer
        _samplesPerCycle = samplesPerBuffer / cyclesPerBuffer

        ' Determine the actual sample clock rate
        timingSubobject.SampleClockRate = _desiredSampleClockRate
        _resultingSampleClockRate = timingSubobject.SampleClockRate

        _resultingFrequency = _resultingSampleClockRate / (samplesPerBuffer / cyclesPerBuffer)

        If (type = WaveformType.SineWave) Then
            _data = GenerateSineWave(_resultingFrequency, amplitude, _resultingSampleClockRate, samplesPerBuffer)
        End If

    End Sub

    Public ReadOnly Property Data() As Double()
        Get
            Return _data
        End Get
    End Property

    Public ReadOnly Property ResultingSampleClockRate() As Double
        Get
            Return _resultingSampleClockRate
        End Get
    End Property

    Public Shared Function GenerateSineWave( _
        ByVal frequency As Double, _
        ByVal amplitude As Double, _
        ByVal sampleClockRate As Double, _
        ByVal samplesPerBuffer As Double) As Double()

        Dim deltaT As Double
        Dim intSamplesPerBuffer As Integer

        deltaT = 1 / sampleClockRate ' sec./samp
        intSamplesPerBuffer = CInt(samplesPerBuffer) - 1

        Dim rVal(intSamplesPerBuffer - 1) As Double

        For i As Integer = 0 To intSamplesPerBuffer - 1
            rVal(i) = amplitude * Math.Sin((2.0 * Math.PI) * frequency * (i * deltaT))
        Next

        Return rVal
    End Function

    Public Shared Sub InitComboBox(ByVal box As System.Windows.Forms.ComboBox)
        Dim obj(0) As Object
        obj(0) = "Sine Wave"

        box.Items.Clear()
        box.Items.AddRange(obj)
        box.Sorted = False
        box.DropDownStyle = ComboBoxStyle.DropDownList
        box.Text = "Sine Wave"
    End Sub

    Private _data As Double()
    Private _resultingSampleClockRate As Double
    Private _resultingFrequency As Double
    Private _desiredSampleClockRate As Double
    Private _samplesPerCycle As Double

End Class

File: NI-DAQ/Examples/DotNET4.0/Analog Out/Generate Voltage/GenMultVoltUpdatesIntClk_DigStart/CS/FunctionGenerator.cs
using System;
using NationalInstruments.DAQmx;
using System.Diagnostics;

namespace NationalInstruments.Examples
{
    public enum WaveformType {
        SineWave     = 0
    }

	public class FunctionGenerator
	{
        public FunctionGenerator(
            Timing timingSubobject,
            string desiredFrequency,
            string samplesPerBuffer,
            string cyclesPerBuffer,
            string type,
            string amplitude)
        {
            WaveformType t = new WaveformType();
            t = WaveformType.SineWave;
            if(type == "Sine Wave")
                t = WaveformType.SineWave;
            else
                Debug.Assert(false,"Invalid Waveform Type");

            Init(
                timingSubobject,
                Double.Parse(desiredFrequency),
                Double.Parse(samplesPerBuffer),
                Double.Parse(cyclesPerBuffer),
                t,
                Double.Parse(amplitude));
        }

        public FunctionGenerator(
            Timing timingSubobject,
            double desiredFrequency,
            double samplesPerBuffer,
            double cyclesPerBuffer,
            WaveformType type,
            double amplitude)
        {
            Init(
                timingSubobject,
                desiredFrequency,
                samplesPerBuffer,
                cyclesPerBuffer,
                type,
                amplitude);
        }

        private void Init(
		    Timing timingSubobject,
            double desiredFrequency,
            double samplesPerBuffer,
            double cyclesPerBuffer,
            WaveformType type,
            double amplitude)
		{
            if(desiredFrequency <= 0)
                throw new ArgumentOutOfRangeException("desiredFrequency",desiredFrequency,"This parameter must be a positive number");
            if(samplesPerBuffer <= 0)
                throw new ArgumentOutOfRangeException("samplesPerBuffer",samplesPerBuffer,"This parameter must be a positive number");
            if(cyclesPerBuffer <= 0)
                throw new ArgumentOutOfRangeException("cyclesPerBuffer",cyclesPerBuffer,"This parameter must be a positive number");

            // First configure the Task timing parameters
            if(timingSubobject.SampleTimingType == SampleTimingType.OnDemand)
                timingSubobject.SampleTimingType = SampleTimingType.SampleClock;

            _desiredSampleClockRate = (desiredFrequency * samplesPerBuffer) / cyclesPerBuffer;
            _samplesPerCycle = samplesPerBuffer / cyclesPerBuffer;

            // Determine the actual sample clock rate
            timingSubobject.SampleClockRate = _desiredSampleClockRate;
            _resultingSampleClockRate = timingSubobject.SampleClockRate;

            _resultingFrequency = _resultingSampleClockRate / (samplesPerBuffer / cyclesPerBuffer);

            switch(type)
            {
                case WaveformType.SineWave:
                    _data = GenerateSineWave(_resultingFrequency,amplitude,_resultingSampleClockRate,samplesPerBuffer);
                    break;
                default:
                    // Invalid type value
                    Debug.Assert(false);
                    break;
            }
        }

        public double[] Data
        {
            get
            {
                return _data;
            }
        }

        public double ResultingSampleClockRate
        {
            get
            {
                return _resultingSampleClockRate;
            }
        }

        public static double[] GenerateSineWave(
            double frequency, 
            double amplitude,
            double sampleClockRate,
            double samplesPerBuffer)
        {
            double deltaT = 1/sampleClockRate; // sec./samp
            int intSamplesPerBuffer = (int)samplesPerBuffer;

            double[] rVal = new double[intSamplesPerBuffer];

            for(int i=0;i<intSamplesPerBuffer;i++)
                rVal[i] = amplitude * Math.Sin( (2.0 * Math.PI) * frequency * (i*deltaT) );

            return rVal;
        }

        public static void InitComboBox(System.Windows.Forms.ComboBox box)
        {
            box.Items.Clear();
            box.Items.AddRange(new object[] {
                "Sine Wave"});
            box.Sorted = false;
            box.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            box.Text = "Sine Wave";
        }

        private double[] _data;
        private double _resultingSampleClockRate;
        private double _resultingFrequency;
        private double _desiredSampleClockRate;
        private double _samplesPerCycle;
    }
}
